lix/source-code/client/src/openRepository.ts

```diff
@@ -183,56 +328,237 @@ export async function openRepository(
-				})
-			})
+			.then(() => checkOutPlaceholders())
 			.finally(() => {
 				pending = undefined
 			})
 			.catch((newError: Error) => {
 				setErrors((previous) => [...(previous || []), newError])
 			})
-
-		await pending
+	} else {
+		console.info("Using existing cloned repo")
 	}
 
 	// delay all fs and repo operations until the repo clone and checkout have finished, this is preparation for the lazy feature
-	function delayedAction({ execute }: { execute: () => any }) {
+	function delayedAction({
+		execute,
+		prop,
+		argumentsList,
+	}: {
+		execute: () => any
+		prop: any
+		argumentsList: any[]
+	}) {
+		const filename = argumentsList?.[0]?.replace(/^(\.)?(\/)?\//, "")
+		const pathParts = filename?.split("/") || []
+		const rootObject = pathParts[0]
+
+		if (
+			experimentalFeatures.lazyClone &&
+			rootObject &&
+			rootObject !== ".git" &&
+			filename !== ".gitignore" &&
+			["readFile", "readlink", "writeFile"].includes(prop) &&
+			!checkedOut.has(rootObject) &&
+			!checkedOut.has(filename)
+		) {
+			if (debug) {
+				console.warn("delayedAction", {
+					prop,
+					argumentsList,
+					rootObject,
+					checkedOut,
+					filename,
+					pending,
+					nextBatch,
+				})
+			}
+			// TODO: optimize writes? only needs adding the head hash to staging instead of full checkout....
+			// if (prop === "writeFile") {
+			// 	checkedOut.add(filename)
+			// } else {
+
+			nextBatch.push(filename)
+
+			// }
+
+			//  && nextBatch.length > 0
+			if (!pending) {
+				pending = doCheckout()
+			}
+		} else {
+			// excluded files execute immediately without waiting for other lazy actions either
+			return execute()
+		}
+
 		if (pending) {
-			return pending.then(execute)
+			return pending.then(execute).finally(() => {
+				if (debug) {
+					console.warn("executed", filename, prop)
+				}
+			})
 		}
 
 		return execute()
 	}
 
+	const nodeishFs = withProxy({
+		nodeishFs: rawFs,
+		verbose: debug,
+		description: "app",
+		intercept: lazyFS ? delayedAction : undefined,
+	})
+
+	const add = (filepath: string | string[]) =>
+		isoGit.add({
+			fs: withProxy({
+				nodeishFs: rawFs,
+				verbose: debug,
+				description: "add",
+			}),
+			parallel: true,
+			dir,
+			cache,
+			filepath,
+		})
+
+	const remove = (filepath: string) =>
+		isoGit.remove({
+			fs: withProxy({
+				nodeishFs: rawFs,
+				verbose: debug,
+				description: "remove",
+			}),
+			dir,
+			cache,
+			filepath,
+		})
+
+	async function emptyWorkdir() {
+		const statusResult = await statusList()
+		if (statusResult.length > 0) {
+			throw new Error("could not pull, uncommitted changes")
+		}
+
+		const listing = (await rawFs.readdir("/")).filter((entry) => {
+			return !checkedOut.has(entry) && entry !== ".git" && entry !== ".gitignore"
+		})
+
+		const notIgnored = (
+			await Promise.all(
+				listing.map((entry) =>
+					isIgnored({ fs: rawFs, dir, filepath: entry }).then((ignored) => {
```

Comment:
@martin-lysk  wow also an amazing catch, how did you even think about this? only solution is to add recursive full repo checkout / materialization of nested .gitignore files on initial placeholder creation same way as i do with the top level gitignore. agree?

Reply:
This makes - i guess this is also not a big issues since we know if we have a nested ignore when we look at the tree object + ignore files are rather small usually